### ISSUE TITLE
Force opening of non-valid XML files

### ISSUE THREAD

**Issue by Azrael3000** [zeux/pugixml#34]
Hi,

first of all let me say thanks for your great code. I am currently using it to read certain VTK files without requiring the complete VTK install. The problem is that in certain cases the vtk file is not valid XML as it will contain a node that contains binary data. E.g.

    <node attribute="1">
      Some binary garbage here
    </node>

pugixml rightfully rejects reading this file. However, I was wondering whether there is a way to force reading of this file. If not, is such a feature planned?

Thanks in advance,
Arno

**Comment by zeux** [comment]
Unfortunately, this seems to be impossible.

Right now pugixml internally relies on the parse buffer being null-terminated. If the source buffer is not, the last character is replaced with \0. This is important for parsing efficiency since it significantly simplifies end-of-stream checks.

However, imagine that this was not the case and pugixml treated 0 as just another character value.

There could be a valid VTK file with a structure like this:

<VTKFile type="..." version="version" byte_order="byte-order" ...>
 <AppendedData encoding="raw">
   _\0\0\0\xf</AppendedData>\0\0\0\xe<AppendedData>
 </AppendedData>
</VTKFile>

E.g. an AppendedData section with two binary chunks with </AppendedData> as the contents of the first chunk and <AppendedData> as the contents of the second one.

If you look at this file as if it were XML (disregarding the null bytes), VTKFile node would have two AppendedData children; if you know the encoding of the AppendedData binary content you'd know to parse the contents separately and have just one AppendedData child.

In other words, it's impossible to parse internal non-XML contents without knowing the encoding format if you want to preserve the outer structure.

**Comment by Azrael3000** [comment]
Yeah my feeling was that this will not work for exactly this reason. I have an idea for a workaround though, so I'm gonna hack that together. Thanks.
